cmake/Sanitizers.cmake

```diff
@@ -0,0 +1,94 @@
+find_program(CLANG_APPLY_EXE NAMES "clang-apply-replacements" DOC "Path to clang-apply-replacements executable")
+find_program(CLANG_TIDY_EXE NAMES "clang-tidy" DOC "Path to clang-tidy executable")
+
+if(CMAKE_GENERATOR MATCHES "Makefile|Ninja")
+    set(USE_SANITIZE_TARGETS TRUE)
+endif()
+
+# Conditional that evaluates to whether or not build optimizations are allowed.
+# This is due to the sanitizers needing the compile commands of each file, not
+# the unity files. CMake will generate files like cmake_pch.cxx, and unity_0.cxx.
+# Both of these files are listed in the compile commands JSON, however, with
+# the current tidy target implementation, the sanitizer cannot know about them.
+# This results in bad/corrupted PCH errors and missing compile commands (errors)
+# for source files included in unity builds.
+if(USE_SANITIZE_TARGETS AND USE_CLANG_TIDY)
+    set(ALLOW_BUILD_OPTIMIZATIONS FALSE)
+else()
+    set(ALLOW_BUILD_OPTIMIZATIONS TRUE)
+endif()
+
+cmake_dependent_option(
+    USE_CLANG_TIDY
+    "Allow the codebase to use the clang-tidy sanitizer"
+    ${ALLOW_BUILD_OPTIMIZATIONS} # Default such that we don't clobber PlasmaTargets.
+    "CLANG_TIDY_EXE"
+    OFF
+)
```

Comment:
Circular dependency:  `USE_CLANG_TIDY` depends on `ALLOW_BUILD_OPTIMIZATIONS`, which is set based on `[...] AND USE_CLANG_TIDY`